src/sharing/ShareWrapper.js

```diff
@@ -0,0 +1,257 @@
+  isSendButtonEnabled: (message: string) => boolean,
+  ownUserId: UserId,
+  sendTo: SendStream | SendPm,
+  sharedData: SharedData,
+|}>;
+
+type State = $ReadOnly<{|
+  message: string,
+  sending: boolean,
+  files: File[],
+|}>;
+
+/**
+ * Wraps Around different sharing screens,
+ * for minimal duplication of code.
+ */
+class ShareWrapper extends React.Component<Props, State> {
+  static contextType = TranslationContext;
+  context: GetText;
+
+  state = (() => {
+    const { sharedData } = this.props;
+    return {
+      message: sharedData.type === 'text' ? sharedData.sharedText : '',
+      sending: false,
+      files: sharedData.type === 'text' ? [] : sharedData.files,
+    };
+  })();
+
+  handleMessageChange = (message: string) => {
+    this.setState({ message });
+  };
+
+  setSending = () => {
+    this.setState({ sending: true });
+  };
+
+  /**
+   * Send received shared data as a message.
+   *
+   * Sends the shared data received from a 3rd party app and possibly modified
+   * by the user, to the server, as a message.
+   */
+  handleSend = async () => {
+    const _ = this.context;
+    const { auth, sendTo, sharedData } = this.props;
+    let messageToSend = this.state.message;
+
+    this.setSending();
+    showToast(_('Sending Message...'));
+    if (sharedData.type === 'file') {
+      const { files } = this.state;
+      for (let i = 0; i < files.length; i++) {
+        const response = await api.uploadFile(auth, files[i].url, files[i].name);
+        messageToSend += `\n[${files[i].name}](${response.uri})\n`;
+      }
+    }
+    const messageData =
+      sendTo.type === 'pm'
+        ? {
+            content: messageToSend,
+            type: 'private',
+            to: JSON.stringify(sendTo.selectedRecipients),
+          }
+        : {
+            content: messageToSend,
+            type: 'stream',
+            subject: sendTo.topic,
+            to: sendTo.stream,
+          };
+
+    try {
+      await api.sendMessage(auth, messageData);
+    } catch (err) {
+      showToast(_('Failed to send message'));
+      logging.error(err);
+      this.onShareTerminated();
+      return;
+    }
+    showToast(_('Message sent'));
+    this.onShareSuccess();
+  };
+
+  onShareTerminated = () => {
+    NavigationService.dispatch(navigateBack());
+  };
+
+  onShareSuccess = () => {
+    const { sendTo } = this.props;
+    switch (sendTo.type) {
+      case 'pm': {
+        const { selectedRecipients } = sendTo;
+        const { ownUserId } = this.props;
+        const recipients = pmKeyRecipientsFromIds(selectedRecipients, ownUserId);
+        const narrow = pmNarrowFromRecipients(recipients);
+        NavigationService.dispatch(replaceWithChat(narrow));
+        break;
+      }
+
+      case 'stream': {
+        const { stream } = sendTo;
+        const narrow = streamNarrow(stream);
+        NavigationService.dispatch(replaceWithChat(narrow));
+        break;
+      }
+
+      default:
+        this.onShareTerminated();
```

Comment:
This case should be impossible, right? It'd be a bug in the app, in fact a bug within the sharing code, if we were to reach this case.

I think treating it as failure is a fine response in the UI, since that's handy. (It'd also be OK to just throw an exception, if there weren't something more specific to do instead that was handy.) But then the other thing we should do is log an error with `logging.error`, so that we see it in Sentry.

Separately, the first thing this case should do is call `ensureUnreachable`. That way Flow will confirm that our types say this case really is impossible.

Here's an example at the end of `handleOutboundEvents.js`:
```js
    default:
      ensureUnreachable(event);
      logging.error(`WebView event of unknown type: ${event.type}`);
      break;
```